lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp

```diff
@@ -257,6 +258,158 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
                       binder.op, resultType, operand);
                   return success();
                 });
+  patterns.onOp(
+      "Loop", 13, [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+        // Get all operands (maxTripCount, cond, ....inits....)
+        llvm::SmallVector<Value> operands;
+        if (binder.tensorOperandsList(operands) || operands.size() == 0) {
+          return rewriter.notifyMatchFailure(binder.op,
+                                             "Failed to get operands");
+        }
+
+        llvm::SmallVector<mlir::Type> operandTypeVec;
+        if (binder.tensorOperandTypes(operandTypeVec) ||
+            operandTypeVec.size() == 0) {
+          return rewriter.notifyMatchFailure(binder.op,
+                                             "Failed to get operandTypes");
+        }
+
+        Region *loopBodyIn;
+        if (binder.getRegionAtIndex(loopBodyIn, 0)) {
+          return rewriter.notifyMatchFailure(binder.op,
+                                             "Failed getting LoopBody Region");
+        }
+
+        // MaxTripCount - tensor int64 scalar (or empty)
+        Value maxTripCountTensor = operands[0];
+        auto maxTripCountInt = rewriter.create<Torch::AtenItemOp>(
+            binder.getLoc(), rewriter.getType<Torch::IntType>(),
+            maxTripCountTensor);
+
+        // Condition - tensor bool scalar (or empty)
+        Value conditionTensor = operands[1];
+        auto conditionInt = rewriter.create<Torch::AtenItemOp>(
+            binder.getLoc(), rewriter.getType<Torch::IntType>(),
+            conditionTensor);
+        auto conditionBool = rewriter.create<Torch::AtenBoolIntOp>(
+            binder.getLoc(), rewriter.getType<Torch::BoolType>(), conditionInt);
+        // To be used for "for like" loop case
+        auto constBoolTrue = rewriter.create<Torch::ConstantBoolOp>(
+            binder.getLoc(), rewriter.getBoolAttr(true));
+
+        // Others (if present) - variadic (can be tensors and scalar values)
+        if (binder.getNumOperands() > 2) {
+          operandTypeVec.erase(operandTypeVec.begin(),
+                               operandTypeVec.begin() + 2);
+          operands.erase(operands.begin(), operands.begin() + 2);
+        }
+
+        auto getOpName = [](Operation *op) -> std::string {
+          std::string name = op->getName().getStringRef().str();
+          if (name != "torch.operator")
+            return name;
+          // for unconverted onnx ops
+          return mlir::dyn_cast<StringAttr>(op->getAttr("name"))
+              .getValue()
+              .str();
+        };
+
+        // PrimLoop Op expectes inputCondition to be boolConstantTrue
+        // to decide if the loopOp is `forlike`. Use loopIsForLike to
+        // ensure appropriate inputCondition is set
+        // Case 1 : loopCondInp -> identity -> terminator(loopCondOut)
+        bool loopIsForLike = false;
+        auto case1ForLike = [&getOpName](Region *loopBody) -> bool {
```

Comment:
Are there other cases?

I see a case 1 but nothing else.

Also, why identity?

I would expect onnx forlikes to have loopCondInp to feed directly to the terminator. Where does the identity come from?

Reply:
Yes, there are other cases. 
Each case is essentially a way of identifying the ways in which one can pass a bool_true as the input condition and how is it handled inside before passing it on to terminator.
Another case is using the same onnx_constant_bool_true, as passed to the onnx.Loop Op,
Another way could be casting an int to bool inside the loop-body and then passing to terminator, and so on..
(did see another way in DLRM's for-loop can't recollect now though)

I've just added a mechanism to catch all such case as we see them in usecases and replace with a single const_bool_true in torch-mlir, hoping that the other part will be removed in DCE in later passes.

> Also, why Identity?

Turns out this was just part of the onnx.Loop example in the documentation. It also seems something like an unsaid rule in onnx.Loop examples to clone all the inputs of Loop Body before usage, hence just picked them up so make sure we cover the (sort of) "native" cases :) Maybe they have their reasons for it which I'm unaware of (clearly)

In this particular [example - loop_11](https://github.com/onnx/onnx/blob/main/docs/Operators.md#examples-84), they clone the condition before passing it on to the terminator op.